Merging row groups throws panic when rows from one group are empty

Merging multiple sorted row groups will throw a panic when one of the groups returns empty rows during `ReadRows`, even if `n` is correctly returned as 0. The reason why we sometimes return empty rows is because we apply filtering during the merge process, and occasionally no rows will match the filter. I think this only happens when the entire row group is empty.

Here is a PR reproducing the issue: https://github.com/parquet-go/parquet-go/pull/67.

Sorry @gernest, I missed your last message. I rebased my branch on latest main and ran the test again. Here is the stack trace from the panic, let me know if this helps:

```
--- FAIL: TestMergeRowGroups (0.01s)
    --- FAIL: TestMergeRowGroups/buffer (0.00s)
        --- FAIL: TestMergeRowGroups/buffer/reproduce_issue_#66,_merging_rows_with_an_empty_row_group (0.00s)
panic: runtime error: index out of range [0] with length 0 [recovered]
	panic: runtime error: index out of range [0] with length 0

goroutine 25 [running]:
testing.tRunner.func1.2({0x177c800, 0xc0000280c0})
	/usr/local/go/src/testing/testing.go:1396 +0x24e
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1399 +0x39f
panic({0x177c800, 0xc0000280c0})
	/usr/local/go/src/runtime/panic.go:884 +0x212
github.com/parquet-go/parquet-go.(*mergeBuffer).read(0xc0022b0000)
	/Users/filippetkovski/Projects/parquet-go/merge.go:463 +0x1ba
github.com/parquet-go/parquet-go.(*mergeBuffer).WriteRowsTo(0xc0022b0000, {0x18d3820, 0xc000515b90})
	/Users/filippetkovski/Projects/parquet-go/merge.go:434 +0x6b
github.com/parquet-go/parquet-go.(*mergedRowGroupRows).WriteRowsTo(0xc000513380, {0x18d3820, 0xc000515b90})
	/Users/filippetkovski/Projects/parquet-go/merge.go:114 +0x1ca
github.com/parquet-go/parquet-go.copyRows({0x18d3820, 0xc000515b90}, {0x276a320, 0xc000513380}, {0x0, 0x0, 0x0})
	/Users/filippetkovski/Projects/parquet-go/row.go:319 +0x3d5
github.com/parquet-go/parquet-go.CopyRows(...)
	/Users/filippetkovski/Projects/parquet-go/row.go:293
github.com/parquet-go/parquet-go_test.copyRowsAndClose({0x18d3820?, 0xc000515b90?}, {0x18df770?, 0xc000513380?})
	/Users/filippetkovski/Projects/parquet-go/parquet_test.go:992 +0xc5
github.com/parquet-go/parquet-go_test.TestMergeRowGroups.func1.1(0xc0004b8ea0)
	/Users/filippetkovski/Projects/parquet-go/merge_test.go:345 +0x505
testing.tRunner(0xc0004b8ea0, 0xc00050eca8)
	/usr/local/go/src/testing/testing.go:1446 +0x10b
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1493 +0x35f
```

--- a/merge.go
+++ b/merge.go
@@ -378,7 +378,6 @@ func (m *mergeBuffer) release() {
 
 func (m *mergeBuffer) fill() error {
 	m.len = len(m.rows)
-	var emptyRowGroups int
 	for i := range m.rows {
 		m.head[i] = 0
 		m.buffer[i] = m.buffer[i][:mergeBufferSize]
@@ -388,18 +387,9 @@ func (m *mergeBuffer) fill() error {
 				return err
 			}
 		}
-		if n == 0 {
-			emptyRowGroups++
-		}
 		m.buffer[i] = m.buffer[i][:n]
 	}
 	heap.Init(m)
-	// heap.Init sorts m.rows pushing empty row groups at the end of the slice.
-	//
-	// We can call heap.Pop(m) n times where n is emptyRowGroups. But it has the same
-	// effect as decrementing m.len while doing extra work. We are avoiding extra
-	// work by directly decrementing m.len by emptyRowGroups.
-	m.len -= emptyRowGroups
 	return nil
 }
 
@@ -470,6 +460,10 @@ func (m *mergeBuffer) left() bool {
 func (m *mergeBuffer) read() (n int64) {
 	for n < int64(len(m.copy)) && m.Len() != 0 {
 		r := m.buffer[:m.len][0]
+		if len(r) == 0 {
+			heap.Pop(m)
+			continue
+		}
 		m.copy[n] = append(m.copy[n][:0], r[m.head[0]]...)
 		m.head[0]++
 		n++

--- a/merge.go
+++ b/merge.go
@@ -378,6 +378,7 @@ func (m *mergeBuffer) release() {
 
 func (m *mergeBuffer) fill() error {
 	m.len = len(m.rows)
+	var emptyRowGroups int
 	for i := range m.rows {
 		m.head[i] = 0
 		m.buffer[i] = m.buffer[i][:mergeBufferSize]
@@ -387,9 +388,18 @@ func (m *mergeBuffer) fill() error {
 				return err
 			}
 		}
+		if n == 0 {
+			emptyRowGroups++
+		}
 		m.buffer[i] = m.buffer[i][:n]
 	}
 	heap.Init(m)
+	// heap.Init sorts m.rows pushing empty row groups at the end of the slice.
+	//
+	// We can call heap.Pop(m) n times where n is emptyRowGroups. But it has the same
+	// effect as decrementing m.len while doing extra work. We are avoiding extra
+	// work by directly decrementing m.len by emptyRowGroups.
+	m.len -= emptyRowGroups
 	return nil
 }
 

--- a/merge_test.go
+++ b/merge_test.go
@@ -155,6 +155,7 @@ func TestMergeRowGroups(t *testing.T) {
 					},
 					Person{FirstName: "Han", LastName: "Solo"},
 				),
+
 				sortedRowGroup(
 					[]parquet.RowGroupOption{
 						parquet.SchemaOf(Person{}),
@@ -165,8 +166,19 @@ func TestMergeRowGroups(t *testing.T) {
 						),
 					},
 				),
+				sortedRowGroup(
+					[]parquet.RowGroupOption{
+						parquet.SortingRowGroupConfig(
+							parquet.SortingColumns(
+								parquet.Ascending("LastName"),
+							),
+						),
+					},
+					Person{FirstName: "Obiwan", LastName: "Kenobi"},
+				),
 			},
 			output: sortedRowGroup(nil,
+				Person{FirstName: "Obiwan", LastName: "Kenobi"},
 				Person{FirstName: "Han", LastName: "Solo"},
 			),
 		},

--- a/merge_test.go
+++ b/merge_test.go
@@ -135,7 +135,41 @@ func TestMergeRowGroups(t *testing.T) {
 				Person{FirstName: "Han", LastName: "Solo"},
 			),
 		},
-
+		{
+			scenario: "reproduce issue #66, merging rows with an empty row group",
+			options: []parquet.RowGroupOption{
+				parquet.SortingRowGroupConfig(
+					parquet.SortingColumns(
+						parquet.Ascending("LastName"),
+					),
+				),
+			},
+			input: []parquet.RowGroup{
+				sortedRowGroup(
+					[]parquet.RowGroupOption{
+						parquet.SortingRowGroupConfig(
+							parquet.SortingColumns(
+								parquet.Ascending("LastName"),
+							),
+						),
+					},
+					Person{FirstName: "Han", LastName: "Solo"},
+				),
+				sortedRowGroup(
+					[]parquet.RowGroupOption{
+						parquet.SchemaOf(Person{}),
+						parquet.SortingRowGroupConfig(
+							parquet.SortingColumns(
+								parquet.Ascending("LastName"),
+							),
+						),
+					},
+				),
+			},
+			output: sortedRowGroup(nil,
+				Person{FirstName: "Han", LastName: "Solo"},
+			),
+		},
 		{
 			scenario: "row groups sorted by descending last name",
 			options: []parquet.RowGroupOption{

--- a/row_group_test.go
+++ b/row_group_test.go
@@ -120,7 +120,18 @@ func fileRowGroup(rowGroup parquet.RowGroup) parquet.RowGroup {
 	if err != nil {
 		panic(err)
 	}
-	return f.RowGroups()[0]
+	g := f.RowGroups()
+	if len(g) > 0 {
+		return g[0]
+	}
+	// There is a test checking for a panic when merging empty row groups. One of
+	// the input is an empty row group which leads to this path.
+	//
+	// It is unnecessary to also return an empty row group here because the
+	// behavior is triggered by custom row group implementation.
+	//
+	// buffer  scenario check  should be sufficient to cover for the issue.
+	return nil
 }
 
 func TestWriteRowGroupClosesRows(t *testing.T) {

